jax/_src/numpy/lax_numpy.py

```diff
@@ -4328,8 +4313,8 @@ def _arange_dynamic(
   gap = step if step_gt_0 else - step
   distance = (stop - start) if step_gt_0 else (start - stop)
   size = core.max_dim(0, distance + gap - 1) // gap
-  return (array(start, dtype=dtype) +
-          array(step, dtype=dtype) * lax.iota(dtype, size))
+  return (array(start, dtype=dtype, device=device) +
+          array(step, dtype=dtype, device=device) * lax.iota(dtype, size))
```

Comment:
`start` and `step` here are scalars, so the sharding likely does not apply to them. We want to apply the sharding to the `iota`.